cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+                          enum cfopaction action,
+                          bool *verified)
+{
+    Log(LOG_LEVEL_DEBUG, "Installing repo type package: %d", is_in_cache);
+    
+    /* Package is not present in cache. */
+    if (is_in_cache == 0)
+    {
+        const char *version = package_info->version;
+        if (package_info->version &&
+                StringSafeEqual(package_info->version, "latest"))
+        {
+            version = NULL;
+        }
+        if (action == cfa_warn || DONTDO)
+        {
+            Log(LOG_LEVEL_NOTICE, "Should install repo type package: %s",
+                package_info->name);
+            return PROMISE_RESULT_FAIL;
+        }
+        
+        *verified = false; /* Verification will be done elsewhere. */
+        return InstallPackage(policy_data->package_options, PACKAGE_TYPE_REPO,
+                              package_info->name, version, package_info->arch,
+                              wrapper);
+    }
+    
+    
+    /* We have some packages matching already installed at this point. */
+    
+    
+    /* We have 'latest' version in policy. */
+    if (package_info->version &&
+                StringSafeEqual(package_info->version, "latest"))
+    {
+        /* This can return more than one latest version if we have packages
+         * with different architectures installed. */
+        Seq *latest_versions = 
+                GetVersionsFromUpdates(ctx, package_info, wrapper);
+        if (!latest_versions)
+        {
+            Log(LOG_LEVEL_VERBOSE, "Package '%s' is already in the latest "
+                "version. Skipping instalation.", package_info->name);
+
+            return PROMISE_RESULT_NOOP;
+        }
+        
+        PromiseResult res = PROMISE_RESULT_NOOP;
+        Buffer *install_buffer = BufferNew();
+        Seq *packages_to_install = SeqNew(1, NULL);
+
+        /* Loop through possible updates. */
+        for (int i = 0; i < SeqLength(latest_versions); i++)
+        {
+            PackageInfo *update_package = SeqAt(latest_versions, i);
+            
+            Log(LOG_LEVEL_DEBUG, "Checking for package '%s' version '%s' in "
+                "available updates", package_info->name,
+                update_package->version);
+            
+            /* Just in case some package managers will report highest possible 
+             * version in updates list instead of removing entry if package is 
+             * already in the latest version. */
+            if (IsPackageInCache(ctx, wrapper, package_info->name,
+                                 update_package->version,
+                                 update_package->arch) == 1)
+            {
+                Log(LOG_LEVEL_VERBOSE, "Package version from updates matches "
+                    "one installed. Skipping package instalation.");
+                res = PromiseResultUpdate(res, PROMISE_RESULT_NOOP);
+                continue;
+            }
+            else
```

Comment:
If IsPackageInCache() returned -1, is it still appropriate to go ahead ?
I think you're assuming it was 0 for this else branch ...